### PR TITLE
Fetch RID list in remote part

### PR DESCRIPTION
**Pull request by BECATRUE**
This is a part of #258.

What I have implemented is:
- Add widgets to select a target date and hour.
- Once any is changed, fetch the corresponding RID list.
- Reconstruct signals in dataviewer app.

image

The figure means that RIDs executed today are 765 to 769.

### CODE REVIEW

File: iquip/apps/dataviewer.py [hunk 1]
```diff
@@ -336,33 +340,58 @@ class _RemotePart(QWidget):
...
-        label: Label for showing the execution time of the experiment.
+        dateEdit: QDateEdit for target date.
+        hourCheckBox: QCheckBox for enabling to select an hour.
+        hourSpinBox: QSpinBox for target hour.
+        ridComboBox: QComboBox for showing the RID list of the selected date (and hour).
 
     Signals:
-        ridEditingFinished(rid): The editingFinished signal of spinbox is emitted.
-          The current spinbox value is given as rid.
+        dateHourChanged(date, hour): The target date and hour are changed. If the hour is not set,
```
Comment (kangz12345) [inline iquip/apps/dataviewer.py]: Let's clarify that the date is in the ISO format.

Comment (kangz12345) [reply iquip/apps/dataviewer.py]: Also, it does not give any hint about the type of `hour` when it is set.

File: iquip/apps/dataviewer.py [hunk 2]
```diff
@@ -336,33 +340,58 @@ class _RemotePart(QWidget):
...
-        layout.addWidget(self.spinbox)
-        layout.addWidget(self.label)
+        layout.addWidget(self.dateEdit)
+        layout.addWidget(self.hourCheckBox)
+        layout.addWidget(self.hourSpinBox)
+        layout.addWidget(self.ridComboBox)
         # signal connection
-        self.spinbox.editingFinished.connect(self._editingFinished)
+        self.dateEdit.dateChanged.connect(self.updateRidComboBox)
+        self.hourCheckBox.stateChanged.connect(self.enableHourSpinBox)
```
Comment (kangz12345) [inline iquip/apps/dataviewer.py]: How about using `clicked(checked)` signal and connect `self.hourComboBox.setEnabled` as its slot? `QCheckBox` inherits `QAbstractButton`, but I'm not sure you can use `clicked` signal.

File: iquip/apps/dataviewer.py [hunk 3]
```diff
@@ -336,33 +340,58 @@ class _RemotePart(QWidget):
...
+        self.hourCheckBox.stateChanged.connect(self.updateRidComboBox)
+        self.hourSpinBox.valueChanged.connect(self.updateRidComboBox)
+
+    @pyqtSlot(int)
+    def enableHourSpinBox(self, state: int):
```
Comment (kangz12345) [inline iquip/apps/dataviewer.py]: In this case, the method name should be `setEnabled...`.

File: iquip/apps/dataviewer.py [hunk 4]
```diff
@@ -886,18 +905,25 @@ def _filter(self, names: List[str]) -> List[str]:
         """
         return [name for name in names if not name.endswith((".parameters", ".units"))]
 
+    def stop(self):
+        """Stops the thread."""
+        try:
+            self.websocket.close()
```
Comment (kangz12345) [inline iquip/apps/dataviewer.py:903]: Is this thread-safe?

**Review by kangz12345 — CHANGES_REQUESTED** [review]
The way of getting the realtime part and remote part looks so weird... Sorry for the messy design 😄 Let's modify that later.

File: iquip/apps/dataviewer.py (hunk 4 above)
Comment (BECATRUE) [reply iquip/apps/dataviewer.py:903]: Umm.. I think it is thread-safe as this method is located in `QThread`.
However I'm not sure about this..

File: iquip/apps/dataviewer.py (hunk 2 above)
Comment (BECATRUE) [reply iquip/apps/dataviewer.py]: Wow, it works well! I'll do that.

File: iquip/apps/dataviewer.py (hunk 3 above)
Comment (BECATRUE) [reply iquip/apps/dataviewer.py]: I used `clicked` signal instaed of `stateChanged` signal, so now this method becomes unusable.

File: iquip/apps/dataviewer.py (hunk 4 above)
Comment (kangz12345) [reply iquip/apps/dataviewer.py:903]: I mean while the `QThread` is communicating with the artiq proxy using `self.websocket`, someone can call `stop()` in another thread.
Then the `websocket` object must support the thread-safe API.
Several different thread can use the `websocket` instance in this code.

Comment (BECATRUE) [reply iquip/apps/dataviewer.py:903]: I see. According to [docs](https://websockets.readthedocs.io/en/stable/reference/sync/client.html#websockets.sync.client.ClientConnection.close), `close()` waits for all incoming messages to be read with `recv()`, so it looks thread-safe.
And I can't say for sure, but looking at the code of `websocket`, it uses `threading.Lock()` for guaranteeing thread-safety.

Comment (kangz12345) [reply iquip/apps/dataviewer.py:903]: Okay, then maybe it is thread-safe!

**Review by kangz12345 — APPROVED** [review]
LGTM

**Pull request merged**